### PR TITLE
[RFC] Avoid dispatching calls to abstract methods

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

Possible solution for https://github.com/sorbet/sorbet/issues/7524.

When a method is marked as `abstract` we shouldn't be able to dispatch calls to it, which is the runtime behavior.

Let's consider this example:

```rb
class A
  def foo(x); end
end

class B < A
  sig { abstract.params(x: Integer, y: Integer).void }
  def foo(x, y); end
end

class C < B
  sig { override.params(x: Integer, y: Integer).void }
  def foo(x, y)
    # Calling `super` in this method actually calls `A#foo` and not `B#foo`, because `B#foo` is marked as abstract
    super()      # so this call should error as we expect only one argument `x`
    super(42)    # while this call should be okay
  end
end
```

[[sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%0A%20%20def%20foo%28x%29%3B%20end%0Aend%0A%0Aclass%20B%20%3C%20A%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20abstract!%0A%0A%20%20sig%20%7B%20abstract.params%28x%3A%20Integer%2C%20y%3A%20Integer%29.void%20%7D%0A%20%20def%20foo%28x%2C%20y%29%3B%20end%0Aend%0A%0Aclass%20C%20%3C%20B%0A%20%20sig%20%7B%20override.params%28x%3A%20Integer%2C%20y%3A%20Integer%29.void%20%7D%0A%20%20def%20foo%28x%2C%20y%29%0A%20%20%20%20%23%20Calling%20%60super%60%20in%20this%20method%20actually%20calls%20%60A%23foo%60%20and%20not%20%60B%23foo%60%2C%20because%20%60B%23foo%60%20is%20marked%20as%20abstract%0A%20%20%20%20super%28%29%20%20%20%20%20%20%23%20so%20this%20call%20should%20error%20as%20we%20expect%20only%20one%20argument%20%60x%60%0A%20%20%20%20super%2842%29%20%20%20%20%23%20while%20this%20call%20should%20be%20okay%0A%20%20end%0Aend)]

At runtime, `sorbet-runtime` will actually remove the definition of `B#foo` so when we call `super` from `C#foo`, the call is actually dispatched to `A#foo`.

This behavior can be confusing when compared to the static type checker behavior that will try to check the arguments to `foo` using the abstract definition.

This is also true when calling initialize:

```rb
class A
  abstract!

  sig { abstract.params(x: Integer, y: Integer).void }
  def initialize(x, y); end
end

class B < A
end

B.new(42, 42) # counterintuitively, this should be an error as the actual `initialize` called is `Object#initialize`
B.new         # similarly this should be okay
```

[[sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%20%20%0A%20%20abstract!%0A%0A%20%20sig%20%7B%20abstract.params%28x%3A%20Integer%2C%20y%3A%20Integer%29.void%20%7D%0A%20%20def%20initialize%28x%2C%20y%29%3B%20end%0Aend%0A%0Aclass%20B%20%3C%20A%0Aend%0A%0AB.new%2842%2C%2042%29%20%23%20counterintuitively%2C%20this%20should%20be%20an%20error%20as%20the%20actual%20%60initialize%60%20called%20is%20%60Object%23initialize%60%0AB.new%20%20%20%20%20%20%20%20%20%23%20similarly%20this%20should%20be%20okay)]

With the proposed patch, the call resolves to the correct method as we can see in the error:

```
test.rb:16 Wrong number of arguments for constructor. Expected: 0, got: 2 https://srb.help/7004
    16 | B.new(42, 42)
               ^^  ^^
    https://github.com/sorbet/sorbet/tree/master/rbi/core/basic_object.rbi#L228: initialize defined here
     228 |  def initialize(); end
            ^^^^^^^^^^^^^^^^
```

While this works, this could be very confusing for the end-user as there is no indication in the error as to why `A#initialize` was skipped during the resolution. Hence the `[RFC]` tag.

### Test plan

No test yet, let's discuss it first.

### CODE REVIEW

**Comment by jez** [comment]
Hey thanks, this was on @neilparikh’s radar and I think he might have been working on a fix. I am happy to look at this first thing when I get back from vacation next week.

**Review by neilparikh — COMMENTED** [review]
Left one question, but this is basically the approach I had in mind, so I don't have any objections to this.

I agree that this is confusing UX, but I think the only alternative is changing sorbet-runtime to match static, which we probably can't do.

One thing that we could try to make the error message more clear is mention which methods were skipped because they were abstract. So in your example, the error for the `super()` call could have an error section like `Method B#foo was skipped because it is abstract`. Of course, it doesn't help explain why `super(42)` doesn't have an error on it.

I think jump to def on the `super` calls would go to `A#foo` (and we should fix it if it doesn't), so that would help explain it a bit as well.

File: core/types/calls.cc [hunk 1]
```diff
@@ -663,6 +663,33 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
...
+    //   sig { override.params(x: Integer, y: Integer).void }
+    //   def foo(x, y)
+    //     # Calling `super` in this method actually calls `A#foo` and not `B#foo`
+    //     super()      # so this call should error as we expect only one argument `x`
+    //     super(42)    # while this should be okay
+    //   end
+    // end
+    while (mayBeOverloaded.exists() && mayBeOverloaded.data(gs)->flags.isAbstract) {
+        mayBeOverloaded = mayBeOverloaded.data(gs)->owner.data(gs)->findParentMethodTransitive(gs, targetName);
+    }
```
[the remark this answers is not in the document]
Comment (Morriar) [reply core/types/calls.cc:691]: Good question. I think it actually applies to all method calls, consider the example from the PR description:

```rb
class A
  abstract!

  sig { abstract.params(x: Integer, y: Integer).void }
  def initialize(x, y); end
end

class B < A
end

B.new(42, 42) # counterintuitively, this should be an error as the actual `initialize` called is `Object#initialize`
B.new         # similarly this should be okay
```

What do you think?

**Comment by neilparikh** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_PTOPLJGnJwnjRH
→ https://go/builds/bui_PTOPGdLddilGHj
→ https://go/builds/bui_PTOPpQGrRNMOGM

**Comment by Morriar** [comment]
Superseded by https://github.com/sorbet/sorbet/pull/7853. Thanks @neilparikh!

**Pull request closed without merging**